proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4209,1003 @@ lemma refillUnblockCheck_corres':
+                  and (\<lambda>s. pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
+                                    (sc_refill_cfgs_of s) sc_ptr)
+                  and (\<lambda>s. pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) \<le> unat max_time)
+                                    (sc_refill_cfgs_of s) sc_ptr))
+                 (valid_refills' sc_ptr)
+                 (head_insufficient_loop sc_ptr)
+                 (headInsufficientLoop scPtr)"
+  apply (clarsimp simp: head_insufficient_loop_def headInsufficientLoop_def runReaderT_def)
+  apply (rule_tac Q="active_sc_at' scPtr" in corres_cross_add_guard)
+   apply (fastforce dest: active_sc_at'_cross)
+  apply (rule corres_whileLoop_abs; simp?)
+       apply (frule head_insufficient_equiv[where scPtr=scPtr]; simp?)
+       apply (fastforce intro: refills_unat_sum_MIN_BUDGET_implies_non_empty_refills)
+      apply (corressimp corres: nonOverlappingMergeRefills_corres)
+      apply (fastforce dest: head_insufficient_length_at_least_two)
+     apply (wpsimp wp: non_overlapping_merge_refills_no_fail)
+     apply (fastforce intro!: refills_unat_sum_MIN_BUDGET_implies_non_empty_refills sc_atD1
+                        simp: obj_at_def)
+    apply (wpsimp wp: non_overlapping_merge_refills_refills_unat_sum_lower_bound
+                      non_overlapping_merge_refills_refills_unat_sum)
+    apply (fastforce dest: head_insufficient_length_greater_than_one)
+   apply (wpsimp wp: nonOverlappingMergeRefills_valid_objs')
+  apply (fastforce intro!: non_overlapping_merge_refills_terminates)
+  done
+
+lemma refillEmpty_sp:
+  "\<lbrace>P\<rbrace>refillEmpty scp \<lbrace>\<lambda>rv s. P s \<and> (\<forall>ko. ko_at' ko scp s \<longrightarrow> rv = (scRefillCount ko = 0))\<rbrace>"
+  apply (wpsimp wp: refillEmpty_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma refillFull_sp:
+  "\<lbrace>P\<rbrace> refillFull scp \<lbrace>\<lambda>rv s. P s \<and> (\<forall>ko. ko_at' ko scp s \<longrightarrow> rv = (scRefillCount ko = scRefillMax ko))\<rbrace>"
+  apply (wpsimp wp: refillFull_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma refillFull_corres:
+  "sc_ptr = scPtr
+   \<Longrightarrow> corres (=) (sc_at sc_ptr and pspace_aligned and pspace_distinct)
+                  (valid_refills' scPtr)
+                  (refill_full sc_ptr)
+                  (refillFull scPtr)"
+  apply (rule_tac Q="sc_at' scPtr" in corres_cross_add_guard)
+   apply (fastforce intro: sc_at_cross)
+  apply (clarsimp simp: refill_full_def refillFull_def)
+  apply (rule corres_split'[rotated 2, OF get_sched_context_sp get_sc_sp'])
+   apply (corressimp corres: get_sc_corres)
+  apply (corressimp corres: corres_return_eq_same)
+  apply (fastforce simp: sc_relation_def obj_at_simps valid_refills'_def opt_map_red)
+  done
+
+end
+
+global_interpretation updateRefillTl: typ_at_all_props' "updateRefillTl scPtr f"
+  by typ_at_props'
```

Comment:
All the `typ_at_all_props'` interpretations in this file could be fairly easily moved into one block if we're happy to move some `crunches` around. 

But maybe we could even move a bunch more to one place. I tried, and I can get the `crunches` and the interpretation working for `refillBudgetCheck` at the end of `KHeap_R.thy`, so I'm pretty sure this is viable. @corlewis do you have any thoughts? Maybe we could create a file of its own for this. Of course this'd be moving a lot of things around, so should be done separately. Should I make a JIRA issue for this?

Reply:
Well, moving to the top of this file before the first `context begin interpretation Arch .`, if possible, would be enough for minimising the number of arch interpretation. I don’t see much benefit in moving further beyond that point...

Reply:
For completeness, I think I agree with both of you. It would be nice to have a self-contained place to put all of the `typ_at_all_props'` interpretations, especially with how crunchable they are, but I don't think there's enough benefit to actually invest the time in moving everything.